contracts/cw3-fixed-multisig/src/contract.rs

```diff
@@ -399,5 +401,826 @@ fn calc_range_start(start_after: Option<HumanAddr>) -> Option<Vec<u8>> {
+            required_weight,
+            voting_period,
+        );
+
+        // Propose
+        let msgs = vec![CosmosMsg::Custom(Empty {})];
+        let proposal = HandleMsg::Propose {
+            title: "Title".to_string(),
+            description: "Description".to_string(),
+            msgs,
+            latest: None,
+        };
+        let res = handle(&mut deps, env, proposal).unwrap();
+
+        // Verify
+        assert_eq!(res, HandleResponse::default());
+    }
+
+    #[test]
+    fn test_vote_works() {
+        let mut deps = mock_dependencies(20, &[]);
+
+        let owner = HumanAddr::from("admin0001");
+        let voter1 = HumanAddr::from("voter0001");
+        let voters = vec![
+            Voter {
+                addr: owner.clone(),
+                weight: 0,
+            },
+            Voter {
+                addr: voter1.clone(),
+                weight: 1,
+            },
+            Voter {
+                addr: HumanAddr::from("voter0002"),
+                weight: 2,
+            },
+            Voter {
+                addr: HumanAddr::from("voter0003"),
+                weight: 3,
+            },
+        ];
+        let required_weight = 3;
+        let voting_period = Duration::Time(2000000);
+
+        let env = mock_env(owner, &[]);
+        setup_test_case(
+            &mut deps,
+            env.clone(),
+            voters,
+            required_weight,
+            voting_period,
+        );
+
+        // Propose
+        let msgs = vec![CosmosMsg::Custom(Empty {})];
+        let proposal = HandleMsg::Propose {
+            title: "Title".to_string(),
+            description: "Description".to_string(),
+            msgs,
+            latest: None,
+        };
+        handle(&mut deps, env.clone(), proposal).unwrap();
+
+        // Query the proposal id
+        let query_msg = QueryMsg::ListProposals {
+            start_after: None,
+            limit: None,
+        };
+        let proposals: ProposalListResponse =
+            from_binary(&query(&deps, query_msg).unwrap()).unwrap();
+        let proposals = &proposals.proposals;
+        assert_eq!(1, proposals.len());
+        let proposal = &proposals[0];
+
+        // owner cannot Vote (again)
+        let vote = HandleMsg::Vote {
+            proposal_id: proposal.id,
+            vote: Vote::Yes,
+        };
+        let res = handle(&mut deps, env, vote);
+
+        // Verify
+        assert!(res.is_err());
+        match res.unwrap_err() {
+            StdError::GenericErr { msg, .. } => assert_eq!(&msg, "Already voted on this proposal"),
+            e => panic!("unexpected error: {}", e),
+        }
+
+        // But voter1 can
+        let env = mock_env(voter1, &[]);
+        let vote = HandleMsg::Vote {
+            proposal_id: proposal.id,
+            vote: Vote::Yes,
+        };
+        let res = handle(&mut deps, env, vote).unwrap();
+
+        // Verify
+        assert_eq!(res, HandleResponse::default());
```

Comment:
And then ensure the vote was updated. Also, turned to passed with enough votes.

A test case that it immediately goes to Passed if the proposed has the required weight alone would be good as well (I think this doesn't work now and please improve the code as well)